cmd/contour/serve.go

```diff
@@ -649,31 +648,32 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 		}
 		log.Printf("informer caches synced")
 
-		var grpcServer *grpc.Server
+		grpcServer := xds.NewServer(registry, ctx.grpcOptions(log)...)
 
 		switch ctx.Config.Server.XDSServerType {
-		case config.ContourServerType:
-			switch ctx.Config.Server.XDSServerVersion {
-			case config.XDSv3:
-				grpcServer = contour_xds_v3.RegisterServer(
-					contour_xds_v3.NewContourServer(log, xdscache.ResourcesOf(resources)...),
-					registry,
-					ctx.grpcOptions(log)...)
-			default:
-				grpcServer = contour_xds_v2.RegisterServer(
-					contour_xds_v2.NewContourServer(log, xdscache.ResourcesOf(resources)...),
-					registry,
-					ctx.grpcOptions(log)...)
-			}
 		case config.EnvoyServerType:
-			switch ctx.Config.Server.XDSServerVersion {
-			case config.XDSv3:
-				log.Fatalf("xDS server type %q not yet implemented for server type %q!", ctx.Config.Server.XDSServerVersion, ctx.Config.Server.XDSServerType)
-			default:
-				grpcServer = contour_xds_v2.RegisterServer(
-					envoy_server_v2.NewServer(context.Background(), snapshotCache, nil),
-					registry,
-					ctx.grpcOptions(log)...)
+			for _, ver := range ctx.Config.Server.XDSServerVersions {
```

Comment:
Quick drive-by suggestion ... it might be cleaner to make this more like `xds.NewContourServer(vers, log, other params)` and let the package deal with the versioning internally.

Reply:
I had that originally, but ran into import cycle errors. 😢